server/channels/jobs/batch_migration_worker.go

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package jobs
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/shared/mlog"
+	"github.com/mattermost/mattermost/server/public/shared/request"
+	"github.com/mattermost/mattermost/server/v8/channels/store"
+)
+
+type BatchMigrationWorkerAppIFace interface {
+	GetClusterStatus() []*model.ClusterInfo
+}
+
+// BatchMigrationWorker processes database migration jobs in batches to help avoid table locks.
+//
+// It uses the jobs infrastructure to ensure only one node in the cluster runs the migration at
+// any given time, avoids running the migration until the cluster is uniform, and automatically
+// resets the migration if the cluster version diverges after starting.
+//
+// In principle, the job infrastructure is overkill for this kind of work, as there's a worker
+// created per migration. There's also complication with edge cases, like having to restar the
+// server in order to retry a failed migration job. Refactoring the job infrastructure is left as
+// a future exercise.
+type BatchMigrationWorker struct {
+	jobServer *JobServer
+	logger    mlog.LoggerIFace
+	store     store.Store
+	app       BatchMigrationWorkerAppIFace
+
+	stop    chan bool
+	stopped chan bool
+	jobs    chan model.Job
+
+	migrationKey       string
+	timeBetweenBatches time.Duration
+	doMigrationBatch   func(model.StringMap, store.Store) (model.StringMap, bool, error)
+}
+
+// MakeBatchMigrationWorker creates a worker to process the given migration batch function.
+func MakeBatchMigrationWorker(jobServer *JobServer, store store.Store, app BatchMigrationWorkerAppIFace, migrationKey string, timeBetweenBatches time.Duration, doMigrationBatch func(model.StringMap, store.Store) (model.StringMap, bool, error)) model.Worker {
```

Comment:
Nit: It makes it slightly more readable to name the variables in the `doMigrationBatch` function `func(jobData model.StringMap, store store.Store)`. I had to look at the function definition to see what actually is `model.StringMap`.